unitex/src/fr/umlv/unitex/tfst/tagging/TaggingModel.java

```diff
@@ -105,27 +150,677 @@ protected void updateModel() {
+	 */		
+	void computeAllPaths( int bfp, int bfs, ArrayList<ArrayList<Integer>> allPaths ){
+		ArrayList<ArrayList<Integer>> lstRemove = new ArrayList<>();
+		
+		computePath(boxes[sortedNodes[bfp]], boxes[sortedNodes[bfs]], new ArrayList<Integer>(), allPaths);
+ 		for (int tmp = 0; tmp < allPaths.size(); tmp++) {
+			allPaths.get(tmp).remove(0);
+			if(!containsUseless(allPaths.get(tmp)))
+				lstRemove.add(allPaths.get(tmp));
+		}
+		for(int i = 0; i < lstRemove.size(); i++)
+			allPaths.remove(lstRemove.get(i));
+		
+		for (ArrayList<Integer> arrayList : allPaths) {
+			for (int i = 0; i < arrayList.size(); i++) {
+				if(taggingStates[sortedNodes[arrayList.get(i)]] != TaggingState.NEUTRAL
+						&& taggingStates[sortedNodes[arrayList.get(i)]] != TaggingState.PREFERRED)
+					setBoxStateInternal(sortedNodes[arrayList.get(i)], TaggingState.TO_CHECK);
+			}
+			
+		}
+	}
+	
+	/** @Aissa
+	 * 
+	 * This function checks if the at least one box of the boxes in arrayList
+	 * is in state "USELESS"
+	 * 
+	 * @param arrayList 
+	 * 			ArrayList of the number of each boxes of the path
+	 * 
+	 * @return  true if at least one box is USELESS
+	 * 			false if not
+	 */			
+	private boolean containsUseless(ArrayList<Integer> arrayList) {
+		for(int i = 0; i < arrayList.size(); i++) {
+			if(taggingStates[sortedNodes[arrayList.get(i)]] == TaggingState.USELESS)
+				return true;
+		}
+		return false;
+	}
+
+	/** This function creates all paths between the first box bfp and the last box bfs 
+	 * Put a path in "current"
+	 * When the path is complete, put it in allPaths
+	 * Then we restart
+	 * 
+	 * @param bfp 
+	 * 			The first box in the path
+	 * 
+	 * @param bfs
+	 * 			The last box in the path
+	 * 
+	 * @param current
+	 * 			The list of boxes of the current path
+	 * 
+	 * @param allPath
+	 * 			The list of all paths between the box bfp and bfs
+	 */	
+	void computePath(GenericGraphBox box, GenericGraphBox last, ArrayList<Integer> currentPath, ArrayList<ArrayList<Integer>> allPaths) {
+		if(box.equals(last))
+			allPaths.add(currentPath);
+		else {
+			currentPath.add(renumber[box.getBoxNumber()]);
+			for (GenericGraphBox next : box.transitions) {
+				computePath(boxes[next.getBoxNumber()], last, new ArrayList<Integer>(currentPath), allPaths);
+			}
+			
+		}
+	}
+	
+	void clearList() {
+		lstTok.clear();
+		lstPath.clear();
+	}
+	
+	void createLstContext() {
+		lstContext.clear();
+		for (String arrayList : lstPath) 
+			lstContext.add(new Context(arrayList));
+		
+		for (Context c : lstContext) {
+			for (int i = 0; i < c.path.size(); i++) {
+				if(getTextBoxe(boxes[sortedNodes[c.path.get(i)]]).equals("<E>")
+						&& (boxes[sortedNodes[c.path.get(i)]]).type != GenericGraphBox.INITIAL
+					&& (boxes[sortedNodes[c.path.get(i)]]).type != GenericGraphBox.FINAL) {
+					if(c.currentBox > i)
+						c.currentBox--;
+					c.path.remove(i);
+				}
+			}
+		}
+	}
+	
+	/** This function checks if the new branch can be add to the automaton or not
```

Comment:
add --> added
also, mention this method marks as `TO_CHECK` the `USELESS` boxes that have become accessible and co-accessible